mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -59,6 +59,166 @@ struct RockLinalgAlignPass
+    // could only mean it is being broadcasted. Hence,
+    // putting zero.
+    if (shapes[i] == 1) {
+      vectorOffsets[i] = resultExpr * 0;
+    } else {
+      // Recording the vector offsets here.
+      vectorOffsets[i] = resultExpr;
+      // There is no point of putting a modulo if the size
+      // is equivalent to that.
+      if (i - 1 >= 0 && shapes[i] != strides[i - 1]) {
+        vectorOffsets[i] = vectorOffsets[i] % strides[i - 1];
+      }
+      vectorOffsets[i] = vectorOffsets[i].floorDiv(strides[i]);
+
+      // The resultExpr has to propagated anyway for
+      // other dimensions where the recording in the above
+      // will do the neccesary checks to remove the modulo
+      if (i - 1 >= 0) {
+        resultExpr = resultExpr % strides[i - 1];
+      }
+      resultExpr = resultExpr.floorDiv(strides[i]);
+    }
+  }
+  return vectorOffsets;
+}
+
+// This function will traverse the operands of the linalg.generic folding
+// the view like ops to the indexing maps and returning the ultimate
+// folded map as well as the root operand.
+LogicalResult
+foldViewLikeOperands(PatternRewriter &rewriter, Value op, AffineMap &foldedMap,
+                     Value &rootOp,
+                     SmallVectorImpl<Operation *> &toBeErasedViewLikeOps) {
+  if (memref::CollapseShapeOp collapseOp =
+          op.getDefiningOp<memref::CollapseShapeOp>()) {
+    SmallVector<AffineExpr, 4> resultExprs;
+    int iDimCount = 0;
+    for (SmallVector<int64_t, 2> groups :
+         collapseOp.getReassociationIndices()) {
+      assert(!groups.empty() && "association indices groups cannot be empty");
+      unsigned groupSize = groups.size();
+      SmallVector<int64_t> suffixProduct(groupSize);
+      // Calculate suffix product for all collapse op source dimension sizes.
+      suffixProduct[groupSize - 1] = 1;
+      for (unsigned i = groupSize - 1; i > 0; i--)
+        suffixProduct[i - 1] =
+            suffixProduct[i] * collapseOp.getSrcType().getDimSize(groups[i]);
+      // Derive the index values along all dimensions of the source
+      // corresponding to the index wrt to collapsed shape op output.
+      SmallVector<AffineExpr, 4> srcIndexExpr = getDelinearizedAffineExpr(
+          suffixProduct, collapseOp.getSrcType().cast<ShapedType>().getShape(),
+          rewriter, iDimCount++);
+      for (unsigned i = 0; i < groupSize; i++) {
+        resultExprs.push_back(srcIndexExpr[i]);
+      }
+    }
+    auto representativeMap = AffineMap::get(
+        /*numDims=*/collapseOp.getType().cast<ShapedType>().getShape().size(),
+        /*numSymbols=*/0, resultExprs, rewriter.getContext());
+    foldedMap = representativeMap.compose(foldedMap);
+    toBeErasedViewLikeOps.push_back(collapseOp);
+    return foldViewLikeOperands(rewriter, collapseOp.getViewSource(), foldedMap,
+                                rootOp, toBeErasedViewLikeOps);
+  }
+  if (memref::ExpandShapeOp expandShapeOp =
```

Comment:
Question: given that both collapse and expand shape use the reassociation expressions concept, couldn't you generalize the collapse shape logic into a function that'll allow you to handle both? I'd rather us not have to trip over this in the future.

But if it's too much bother, no problem here.

Reply:
Done. (Good thing I looked at this on a monday morning rather than a friday evening.. :) )